client/client.go

```diff
@@ -129,154 +140,130 @@ func (a *APIError) Unwrap() error {
+	// authorization code to use for requests
+	auth AuthToken
+	// server to use for requests
+	serverURL ServerURL
+	// anonymous function to initiate reauthorization
+	revoke ClientReauthorizeFunc
+}
+
 // function type for reauthorization closure (see func reauthorize@mender.go)
-type ClientReauthorizeFunc func(string) (AuthToken, error)
+type ClientReauthorizeFunc func() (AuthToken, ServerURL, error)
 
-// function type for setting server (in case of multiple fallover servers)
-type ServerManagementFunc func() *MenderServer
+// Reconstruct the given request, taking JWT token and serverURL into account.
+func (c *ReauthorizingClient) reconstructRequest(req *http.Request) (*http.Request, error) {
+	if c.serverURL == "" {
+		return nil, ErrClientUnauthorized
+	}
 
-// Return a new ApiRequest
-func (a *ApiClient) Request(
-	code AuthToken,
-	nextServerIterator ServerManagementFunc,
-	reauth ClientReauthorizeFunc,
-) *ApiRequest {
-	return &ApiRequest{
-		api:                a,
-		auth:               code,
-		nextServerIterator: nextServerIterator,
-		revoke:             reauth,
+	serverURL, err := url.Parse(string(c.serverURL))
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not parse ServerURL from auth manager")
 	}
-}
 
-// ApiRequester compatible helper. The helper can be used for executing API
-// requests that require authorization as provided Do() method will automatically
-// setup authorization information in the request.
-type ApiRequest struct {
-	api *ApiClient
-	// authorization code to use for requests
-	auth AuthToken
-	// anonymous function to initiate reauthorization
-	revoke ClientReauthorizeFunc
-	// anonymous function to set server
-	nextServerIterator ServerManagementFunc
-}
-
-// tryDo is a wrapper around http.Do that also tries to reauthorize
-// on a 401 response (Unauthorized).
-func (ar *ApiRequest) tryDo(req *http.Request, serverURL string) (*http.Response, error) {
-	r, err := ar.api.Do(req)
-	if err == nil && r.StatusCode == http.StatusUnauthorized {
-		// invalid JWT; most likely the token is expired:
-		// Try to refresh it and reattempt sending the request
-		log.Info("Device unauthorized; attempting reauthorization")
-		if jwt, e := ar.revoke(serverURL); e == nil {
-			// retry API request with new JWT token
-			ar.auth = jwt
-			// check if request had a body
-			// (GetBody is optional, and nil if body is empty)
-			if req.GetBody != nil {
-				if body, e := req.GetBody(); e == nil {
-					req.Body = body
-				}
-			}
-			req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", ar.auth))
-			r, err = ar.api.Do(req)
-		} else {
-			log.Warnf("Reauthorization failed with error: %s", e.Error())
+	// First prefill newURL with existing URL.
+	newURL, _ := url.Parse(req.URL.String())
+
+	// Then selectively fill in the prefix from ServerURL.
+	newURL.Scheme = serverURL.Scheme
+	newURL.Host = serverURL.Host
+	newURL.Path = strings.TrimRight(serverURL.Path, "/") +
+		"/" +
+		strings.TrimLeft(req.URL.Path, "/")
```

Comment:
Why we need the first prefill here?

Reply:
The Golang docs state "relative paths may omit leading slash". I have no idea under what circumstances this can happen, and if it is in fact applicable here. But I thought it's better to be safe than sorry, and just remove and readd it every time.

For some context: This took me ages to debug, I was getting successful connections, but it simply refused to use POST, always using GET. It turns out that what happened was that it was connecting to an endpoint with an extra slash in the beginning, and this was being automatically redirected by the Golang server using 301. But 301 redirects do not preserve the method, hence it resorted to GET. This all happened behind my back, outside of my defined handlers, so it took quite some time to find it.